browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -56,16 +84,89 @@ void OnShouldBlockAdResult(const ResponseCallback& next_callback,
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
   if (ctx->blocked_by == kAdBlocked) {
     brave_shields::DispatchBlockedEvent(
-        ctx->request_url,
-        ctx->render_frame_id, ctx->render_process_id, ctx->frame_tree_node_id,
-        brave_shields::kAds);
+        ctx->request_url, ctx->render_frame_id, ctx->render_process_id,
+        ctx->frame_tree_node_id, brave_shields::kAds);
   }
   next_callback.Run();
 }
 
-void OnBeforeURLRequestAdBlockTP(
+void ShouldBlockAdWithOptionalCname(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
     const ResponseCallback& next_callback,
-    std::shared_ptr<BraveRequestInfo> ctx) {
+    std::shared_ptr<BraveRequestInfo> ctx,
+    const base::Optional<std::string> cname) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  task_runner->PostTaskAndReply(
+      FROM_HERE, base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx, cname),
+      base::BindOnce(&OnShouldBlockAdResult, next_callback, ctx));
+}
+
+class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
+ private:
+  mojo::Receiver<network::mojom::ResolveHostClient> receiver_{this};
+  base::OnceCallback<void(base::Optional<std::string>)> cb_;
+
+ public:
+  AdblockCnameResolveHostClient(
+      const ResponseCallback& next_callback,
+      scoped_refptr<base::SequencedTaskRunner> task_runner,
+      std::shared_ptr<BraveRequestInfo> ctx) {
+    cb_ = base::BindOnce(&ShouldBlockAdWithOptionalCname, task_runner,
+                         std::move(next_callback), ctx);
+
+    auto* web_contents = GetWebContents(
+        ctx->render_process_id, ctx->render_frame_id, ctx->frame_tree_node_id);
+    DCHECK(web_contents);
+
+    content::BrowserContext* context = web_contents->GetBrowserContext();
+
+    const auto network_isolation_key = ctx->network_isolation_key;
+
+    network::mojom::ResolveHostParametersPtr optional_parameters =
+        network::mojom::ResolveHostParameters::New();
+    optional_parameters->include_canonical_name = true;
+
+    network::mojom::NetworkContext* network_context =
+        content::BrowserContext::GetDefaultStoragePartition(context)
+            ->GetNetworkContext();
+    network_context->ResolveHost(
+        net::HostPortPair::FromURL(ctx->request_url), network_isolation_key,
+        std::move(optional_parameters), receiver_.BindNewPipeAndPassRemote());
+
+    receiver_.set_disconnect_handler(
+        base::BindOnce(&AdblockCnameResolveHostClient::OnComplete,
+                       base::Unretained(this), net::ERR_NAME_NOT_RESOLVED,
+                       net::ResolveErrorInfo(net::ERR_FAILED), base::nullopt));
+  }
+
+  void OnComplete(
+      int32_t result,
+      const net::ResolveErrorInfo& resolve_error_info,
+      const base::Optional<net::AddressList>& resolved_addresses) override {
+    if (result == net::OK && resolved_addresses) {
+      DCHECK(resolved_addresses.has_value() && !resolved_addresses->empty());
+      std::move(cb_).Run(
```

Comment:
I think we can replace this with a direct call
```
  task_runner->PostTaskAndReply(
      FROM_HERE, base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx, cname),
      base::BindOnce(&OnShouldBlockAdResult, next_callback, ctx));
```
This way,  you don't need to bind and store `cb_`, and `ShouldBlockAdWithOptionalCname` can also be eliminated. `task_runner()` can also be obtained right here. So you would need to only save `next_callback` in constructor

Reply:
@bridiver said he wanted the task runner passed in, perhaps the two of you should discuss?